platform/kube/register.go

```diff
@@ -0,0 +1,196 @@
+func samePorts(ep []v1.EndpointPort, portsMap map[int32]bool) bool {
+	if len(ep) != len(portsMap) {
+		return false
+	}
+	for _, e := range ep {
+		if !portsMap[e.Port] {
+			return false
+		}
+	}
+	return true
+}
+
+// splitEqual splits key=value string into key,value. if no = is found
+// the whole string is the key and value is empty.
+func splitEqual(str string) (string, string) {
+	idx := strings.Index(str, "=")
+	var k string
+	var v string
+	if idx >= 0 {
+		k = str[:idx]
+		v = str[idx+1:]
+	} else {
+		k = str
+	}
+	return k, v
+}
+
+// addLabelsAndAnnotations adds labels and annotations to an object.
+func addLabelsAndAnnotations(obj *meta_v1.ObjectMeta, labels []string, annotations []string) {
+	if obj.Labels == nil {
+		obj.Labels = make(map[string]string, len(labels))
+	}
+	for _, l := range labels {
+		k, v := splitEqual(l)
+		obj.Labels[k] = v
+	}
+	if obj.Annotations == nil {
+		obj.Annotations = make(map[string]string, len(annotations))
+	}
+	for _, a := range annotations {
+		k, v := splitEqual(a)
+		obj.Annotations[k] = v
+	}
+}
+
+// RegisterEndpoint registers the endpoint (and the service if it doesn't
+// already exists). It creates or updates as needed. When creating it adds the
+// optional labels.
+func RegisterEndpoint(client kubernetes.Interface, namespace string, svcName string,
+	ip string, portsList []NamedPort, labels []string, annotations []string) error {
+	getOpt := meta_v1.GetOptions{IncludeUninitialized: true}
+	_, err := client.Core().Services(namespace).Get(svcName, getOpt)
+	if err != nil {
+		glog.Warningf("Got '%v' looking up svc '%s' in namespace '%s', attempting to create it", err, svcName, namespace)
+		svc := v1.Service{}
+		svc.Name = svcName
+		for _, p := range portsList {
+			svc.Spec.Ports = append(svc.Spec.Ports, v1.ServicePort{Name: p.Name, Port: p.Port})
+		}
+		addLabelsAndAnnotations(&svc.ObjectMeta, labels, annotations)
+		_, err = client.CoreV1().Services(namespace).Create(&svc)
+		if err != nil {
+			glog.Error("Unable to create service: ", err)
+			return err
+		}
+	}
+	eps, err := client.CoreV1().Endpoints(namespace).Get(svcName, getOpt)
+	if err != nil {
+		glog.Warningf("Got '%v' looking up endpoints for '%s' in namespace '%s', attempting to create them",
+			err, svcName, namespace)
+		endP := v1.Endpoints{}
+		endP.Name = svcName // same but does it need to be
+		addLabelsAndAnnotations(&endP.ObjectMeta, labels, annotations)
+		eps, err = client.CoreV1().Endpoints(namespace).Create(&endP)
+		if err != nil {
+			glog.Error("Unable to create endpoint: ", err)
+			return err
+		}
+	}
+	// To check equality:
+	portsMap := make(map[int32]bool, len(portsList))
+	for _, e := range portsList {
+		portsMap[e.Port] = true
+	}
+
+	glog.V(2).Infof("Before: found endpoints %+v", eps)
+	matchingSubset := 0
+	for i, ss := range eps.Subsets {
+		glog.V(1).Infof("On ports %+v", ss.Ports)
+		for _, ip := range ss.Addresses {
+			glog.V(1).Infof("Found %+v", ip)
```

Comment:
and here.. this entire loop is just for printing ports. I understand the need for debugging, but on that same vein, it is important to have useful (to everyone) debugging messages :)

Reply:
same

Reply:
read line 162 onward ?